chainer/variable.py

```diff
@@ -283,22 +283,29 @@ def addgrad(self, var):
         src = var._grad
         dst = self._grad
         if src is None:
-            raise ValueError('Source gradient is not set.')
-        if dst is None:
-            raise ValueError('Target gradient is not set.')
+            return
 
-        xp = cuda.get_array_module(dst)
-        if xp is numpy:
-            dst += cuda.to_cpu(src)
-        elif isinstance(src, numpy.ndarray):
-            dst += cuda.to_gpu(src, device=dst)
-        else:
-            dst_dev = dst.device
-            if dst_dev == src.device:
-                dst += src
+        src_dev = cuda.get_device(src)
+        dst_dev = cuda.get_device(self.data)
+
+        if src_dev.id == dst_dev.id:
+            if dst is None:
+                xp = cuda.get_array_module(src)
+                self._grad = xp.copy(src)
```

Comment:
Switch the device here appropriately. Note that cupy.copy allocates an array on the current device (instead of the device that `src` reside). See also the next comment.